src/ui/ui_i18n.js

```diff
@@ -0,0 +1,126 @@
+const I18n =  class {
+  constructor(translations) {
+    this.translations = translations;
+  }
+
+  translate(key) {
+    if (!this.translations.hasOwnProperty(key)) {
+      return null;
+    }
+    return this.translations[key];
+  }
+};
+
+const getLocaleTranslations = async function (locale, server) {
+  const translations = await server.plugins.i18n.getRegisteredLocaleTranslations(locale);
+  return translations;
+};
+
+const getTranslationLocale = function (acceptLanguages, defaultLocale, server) {
+  let localeStr = '';
+
+  if (acceptLanguages === null || acceptLanguages.length <= 0) {
+    return defaultLocale;
+  }
+  const registeredLocales = server.plugins.i18n.getRegisteredTranslationLocales();
+  localeStr = getTranslationLocaleExactMatch(acceptLanguages, registeredLocales);
+  if (localeStr != null) {
+    return localeStr;
+  }
+  localeStr = getTranslationLocaleBestCaseMatch(acceptLanguages, registeredLocales, defaultLocale);
+  return localeStr;
+};
+
+const updateMissingTranslations = async function (defaultLocale, translations, server) {
+  const defaultLocaleTranslations = await getLocaleTranslations(defaultLocale, server);
+  if (defaultLocaleTranslations === null || defaultLocaleTranslations.length <= 0) {
+    return translations;
+  }
+
+  const transKeys = Object.keys(translations).sort();
+  const defaultLocaleTransKeys = Object.keys(defaultLocaleTranslations).sort();
+  if (transKeys === defaultLocaleTransKeys) {
+    return translations;
+  }
+
+  let updatedTranslations = [];
+  for (let key in defaultLocaleTranslations) {
+    if (!defaultLocaleTranslations.hasOwnProperty(key)) continue;
+    if (!translations.hasOwnProperty(key)) {
+      updatedTranslations[key] = defaultLocaleTranslations[key];
+    } else {
+      updatedTranslations[key] = translations[key];
+    }
+  }
+  return updatedTranslations;
+};
```

Comment:
Instead of all this, I would add a method to the core i18n module called `getDefaultTranslations`. Then in `src/ui/index.js` you could accomplish the same thing by doing:

```
const translations = await i18n.getRegisteredLocaleTranslations(locales);
const defaultTranslations = await i18n.getDefaultTranslations();
const fullTranslations = _.defaults(translations, defaultTranslations);
```

This would also negate the need to get the default locale in a try/catch block inside `src/ui/index.js`